torchtext/transforms.py

```diff
@@ -189,3 +193,350 @@ def forward(self, input: Any) -> Any:
+        """
+        encoded: List[str] = []
+        if torch.jit.is_scripting():
+            for b in token:
+                encoded.append(self.byte_encoder[ord(b)])
+        else:
+            encoded = [self.byte_encoder[b] for b in token.encode('utf-8')]
+        return encoded
+
+    @torch.jit.export
+    def regex(self, text: str) -> List[str]:
+        r"""Return a list of tokens, split by regular expression(pcre).
+
+        's|'t|'re|'ve|'m|'ll|'d| ?\p{L}+| ?\p{N}+| ?[^\s\p{L}\p{N}]+|\s+(?!\S)|\s+
+        """
+        return torch.ops.torchtext.gpt2_bpe_pre_tokenizer(text)
+
+    @torch.jit.export
+    def encode(self, text: str) -> List[int]:
+        """Encode text into a list of bpe token ids.
+
+        Split text into a list of token unit, and generate a list of bpe tokens
+        for each token unit. Lastly encode bpe tokens into bpe token ids.
+
+        Args:
+            text: An input text string.
+
+        Returns:
+            A list of bpe token ids represents each bpe tokens.
+
+        For example: "awesome,awe"
+            --> tokenize(regex) --> tokens: ["awesome", ",", "awe"]
+            --> bpe --> bpe tokens: ["aw", "esome"], [","], ["aw", e]
+            --> bpe encode --> bpe token ids: [707, 5927], [11], [707, 68]
+            --> result --> [707, 5927, 11, 707, 68]
+        """
+
+        bpe_token_ids: List[int] = []
+        for token in self.regex(text):
+            for bpe_token in self.bpe(self.byte_encode(token)):
+                bpe_token_ids.append(self.bpe_encoder[bpe_token])
+        return bpe_token_ids
+
+    @torch.jit.export
+    def tokenize(self, text: str) -> List[str]:
+        """Encode text into a list of Token(token_id, start_idx, end_idx)
+
+        Args:
+            text: An input text string.
+
+        Returns:
+            A list of bpe token ids represents each bpe tokens
+
+        For example: "awesome,awe"
+            --> bpe --> bpe tokens: ["aw", "esome"], [","], ["aw", e]
+            --> bpe encode --> bpe token ids: [707, 5927, 11, 707, 68]
+        """
+        bpe_token_ids: List[int] = self.encode(text)
+        bpe_tokens: List[str] = []
+
+        for bpe_token_id in bpe_token_ids:
+            bpe_tokens.append(str(bpe_token_id))
+
+        return bpe_tokens
+
+    def forward(self, input: Any) -> Any:
+        """
+        :param input: Input sentence or list of sentences on which to apply tokenizer.
+        :type input: Union[str, List[str]]
+        :return: tokenized text
+        :rtype: Union[List[str], List[List(str)]]
+        """
+        if torch.jit.isinstance(input, List[str]):
+            tokens: List[List[str]] = []
+            for text in input:
+                tokens.append(self.tokenize(text))
+            return tokens
+        elif torch.jit.isinstance(input, str):
+            return self.tokenize(input)
+        else:
+            raise TypeError("Input type not supported")
```

Comment:
nit: It is more common to perform the input validation before proceeding to the actual logic, but this method seems to be jit-compatible, performance-sensitive and simple, so I guess it's okay.

Reply:
@mthrok Yes, that was my concern as well. However, for now I was sticking to the same interface as the SentencePiece tokenizer. We have been talking about standardizing the Tokenizer interface and we shall bring this point up in that discussion.

Reply:
These type checks are special once that also do type refinement. We could do the input validation, but I guess that would be redundant in this special case of torch-script implementation where we are doing it via type-refinement when we actually need it.

Reply:
Marking this as resolved. We will revisit this as we design the Tokenizers API.